### ISSUE TITLE
Can we make a release incorporating the latest lambdasoup version?

### ISSUE THREAD

**Issue by jbhoot** [PataphysicalSociety/soupault#67]
[The latest release](https://github.com/aantron/lambdasoup/releases/tag/1.1.0) of lambdasoup now supports `:has()` selector.

Is incorporating it in soupault only a matter of updating the lambdasoup dependency?

Happy to make a PR with your guidance.

**Comment by dmbaturin** [comment]
Yes, I saw the release of lambdasoup and was waiting for it to appear in opam. I want to add `HTML.is_text` and `HTML.is_document` to the API, that was impossible until aantron/lambdasoup/pull/52.

If you want to try your hand at it, you can follow the example of [`HTML.is_element`](https://github.com/PataphysicalSociety/soupault/blob/main/src/plugin_api.ml#L167-L171). 

It looks like those those functions work well with nodes coerced to "general nodes":

```
utop # Soup.create_text "test!" |> Soup.coerce |> Soup.is_text ;;
- : bool = true

utop # Soup.create_text "test!" |> Soup.coerce |> Soup.is_document ;;
- : bool = false

utop # Soup.create_soup () |> Soup.coerce |> Soup.is_document ;;
- : bool = true
```

Thus it should be fine to just use [`to_general`](https://github.com/PataphysicalSociety/soupault/blob/main/src/plugin_api.ml#L173) to coerce them and do the respective test.

In fact, I wonder if `HTML.is_element` can be safely rewritten without a special case for `ElementNode`, or if the wrapper type is still necessary at all now that `is_document` is a built-in test. But it's a question for a bigger future release, I suppose.

**Comment by justbhoot** [comment]
Right. Let me update the dependency first, then I will help in incorporating the newer functions too.

**Comment by justbhoot** [comment]
@dmbaturin Are you fine with making an release that only contains the updated dep on lambdasoup?

Or do you want to make a single release *after* incorporating the new functions?

**Comment by dmbaturin** [comment]
Since making a release involves building binaries for all platforms, I'd rather bundle those changes.

**Comment by justbhoot** [comment]
Agreed. Less grunt work that way. I will see if I can take it up tomorrow or in the next week.

**Comment by jbhoot** [comment]
## Regarding HTML.is_text implementation

Based on `HTML.is_element` as you suggested, I wrote the following function:

```
let is_text n =
  match n with
  | GeneralNode n -> Soup.is_text n
  | _ -> false
```

But, from what you wrote:

> Thus it should be fine to just use [to_general](https://github.com/PataphysicalSociety/soupault/blob/main/src/plugin_api.ml#L173) to coerce them and do the respective test.

it seems that you want an implementation like:

```
  let is_text n =
    n |> to_general |> Soup.is_text 
```

Comments?

**Comment by jbhoot** [comment]
## Regarding is_document function

`src/plugin_api.ml` already has an `is_document` function:

```
  let is_document n =
    match n with
    | SoupNode _ -> true
    | _ -> false
```

Perhaps you want to rewrite it to the following?

```
  let is_document n =
    n |> to_general |> Soup.is_document
```

**Comment by dmbaturin** [comment]
@jbhoot Sorry for the late reply, I've been a bit overwhelmed. I merged the PR, I think the change is reasonable and any concerns that made me write it that way no longer apply (if they even existed at all).

Regarding the new release, aantron/lambdasoup/pull/61 was merged today and that may be relevant for some people. Let's see if @aantron plans to make a new release soon.

**Comment by aantron** [comment]
The `:has` selector was already released in Lambda Soup 1.1.0 (as, I understand, you spotted). And yes, I will do a new release now with latest PR.

**Comment by dmbaturin** [comment]
It's official now, thanks for the help and cooperation!

https://github.com/PataphysicalSociety/soupault/releases/tag/4.11.0
